Make RemoveFormat command to work with mixed editable

Ahmad-S792 opened the pull request:
#### 919c3b81c0474dfbe49620a512bdfad1c0f23d0f
<pre>
Make RemoveFormat command to work with mixed editable

Make RemoveFormat command to work with mixed editable
<a href="https://bugs.webkit.org/show_bug.cgi?id=202901">https://bugs.webkit.org/show_bug.cgi?id=202901</a>

Reviewed by Ryosuke Niwa.

Merge - <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=177448

This patch changes assertion in |ApplyStyleCommand::removeInlineStyle()| to
consider |PositionIsBeforeChildren| as same as |PositionIsBeforeAnchor| and
|offsetInContainerNode &lt;= 0|.

In the attached test case, |firstPositionInOrBeforeNode(next.get())| returns
|PositionIsBeforeChildren| for B element. When we don&apos;t have uneditable
element, inline styles are removed by |removeInlineStyleFromElement| for
BODY element rather than each child in BODY element. Webkit doesn&apos;t go the path
contains the assertion.

* Source/WebCore/editing/ApplyStyleCommand.cpp:
(ApplyStyleCommand::removeInlineStyle): Update Assertion
* LayoutTests/editing/execCommand/remove-format-mixed-editable-crash.html: Add Test Case
* LayoutTests/editing/execCommand/remove-format-mixed-editable-crash-expected.txt: Add Test Case Expectation

Canonical link: <a href="https://commits.webkit.org/257399@main">https://commits.webkit.org/257399@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3a023bba658d3fea2e1a1c06c195a687b2949d6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98708 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7918 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31835 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108127 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168379 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102659 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8435 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85287 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91240 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106046 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104390 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6383 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89948 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33385 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88209 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21291 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76311 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1840 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22820 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1749 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45306 "Found 2 new test failures: fast/forms/select-max-length.html, media/video-inactive-playback.html (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5091 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6694 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42275 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3141 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/ApplyStyleCommand.cpp:
```diff
@@ -1,5 +1,6 @@
 /*
- * Copyright (C) 2005, 2006, 2008, 2009 Apple Inc. All rights reserved.
+ * Copyright (C) 2005-2022 Apple Inc. All rights reserved.
+ * Copyright (C) 2022 Google Inc. All rights reserved.
```
You mean 2014? The patch was authored in 2014, right?

Ahmad-S792 replied:
Update - My thought process was that if we are merging in 2022 then even if it is old patch then we should use 2022 but now updated. Thanks!

rniwa commented on LayoutTests/editing/execCommand/remove-format-mixed-editable-crash.html:
```diff
@@ -0,0 +1,18 @@
+onload = function() {
+    document.execCommand('SelectAll');
+    document.execCommand('RemoveFormat');
+    document.body.textContent = 'PASS if Webkit doesn\'t crash.';
```
I don't think it's accurate to say doesn't "crash" since we're talking about debug assertion here.
I suggest we rephrase it as PASS if WebKit doesn't hit any debug assertions.

Ahmad-S792 replied:
Updated to just mention "PASS". Thanks!

Ahmad-S792 committed:
Make RemoveFormat command to work with mixed editable

Make RemoveFormat command to work with mixed editable
https://bugs.webkit.org/show_bug.cgi?id=202901

Reviewed by Ryosuke Niwa.

Merge - https://src.chromium.org/viewvc/blink?view=revision&revision=177448

This patch changes assertion in |ApplyStyleCommand::removeInlineStyle()| to
consider |PositionIsBeforeChildren| as same as |PositionIsBeforeAnchor| and
|offsetInContainerNode <= 0|.

In the attached test case, |firstPositionInOrBeforeNode(next.get())| returns
|PositionIsBeforeChildren| for B element. When we don't have uneditable
element, inline styles are removed by |removeInlineStyleFromElement| for
BODY element rather than each child in BODY element. Webkit doesn't go the path
contains the assertion.

* Source/WebCore/editing/ApplyStyleCommand.cpp:
(ApplyStyleCommand::removeInlineStyle): Update Assertion
* LayoutTests/editing/execCommand/remove-format-mixed-editable-crash.html: Add Test Case
* LayoutTests/editing/execCommand/remove-format-mixed-editable-crash-expected.txt: Add Test Case Expectation

Canonical link: https://commits.webkit.org/257399@main
